File: clients/ts/src/subsumption.ts
/**
 * Port of subsumption.py's `envelope_subsumes` — the skill-delegation
 * Z3 primitive. Never calls `(get-model)`: a counterexample's CONTENT
 * (the concrete step Z3 finds) only ever feeds `reason` strings, which are
 * informative (not compared) — only sat/unsat/unknown drives the verdict.
 */

import type { Envelope, Permission } from "./models.js";
import { SHELL_INTERPRETERS } from "./models.js";
import { Scope, newEnv, compileScalar, smtStr } from "./predicateZ3.js";
import { z3CheckSat, VerificationTimeout } from "./z3client.js";

export interface SubsumptionResult {
  holds: boolean;
  unverifiedInvariants: string[];
  reasons: string[];
}

function globToSmt(pathVar: string, glob: string): string {
  if (glob === "**") return "true";
  if (glob.endsWith("/**")) {
    const prefix = glob.slice(0, -3);
    return `(or (str.prefixof ${smtStr(prefix + "/")} ${pathVar}) (= ${pathVar} ${smtStr(prefix)}))`;
  }
  if (!glob.includes("*")) {
    return `(= ${pathVar} ${smtStr(glob)})`;
  }
  if (glob.startsWith("*") && !glob.slice(1).includes("*")) {
    const suffix = glob.slice(1);
    return `(str.suffixof ${smtStr(suffix)} ${pathVar})`;
  }
  if ((glob.match(/\*/g) ?? []).length === 1) {
    const idx = glob.indexOf("*");
    const prefix = glob.slice(0, idx);
    const suffix = glob.slice(idx + 1);
    return `(and (str.prefixof ${smtStr(prefix)} ${pathVar}) (str.suffixof ${smtStr(suffix)} ${pathVar}))`;
  }
  return "true"; // unsupported shape — approximated permissive (matches oracle)
}

function globUnsupported(glob: string): boolean {
  if (glob === "**" || glob.endsWith("/**") || !glob.includes("*")) return false;
  if (glob.startsWith("*") && !glob.slice(1).includes("*")) return false;
  return (glob.match(/\*/g) ?? []).length !== 1;
}

async function patternsSubsume(
  innerPatterns: string[],
  outerPatterns: string[],
  label: string,
  timeoutMs: number
): Promise<string | null> {
  if (innerPatterns.length === 0) return null;
  const badOuter = outerPatterns.filter(globUnsupported);
  if (badOuter.length > 0) {
    return `${label}: outer declares a glob shape that cannot be soundly encoded (${JSON.stringify(
      badOuter
    )}); cannot prove subsumption → denied`;
  }
  const innerOk =
    innerPatterns.length === 1
      ? globToSmt("v", innerPatterns[0]!)
      : `(or ${innerPatterns.map((g) => globToSmt("v", g)).join(" ")})`;
  const outerOk =
    outerPatterns.length === 0
      ? "false"
      : outerPatterns.length === 1
        ? globToSmt("v", outerPatterns[0]!)
        : `(or ${outerPatterns.map((g) => globToSmt("v", g)).join(" ")})`;
  const script = `(declare-const v String)\n(assert ${innerOk})\n(assert (not ${outerOk}))`;
  const result = await z3CheckSat(script, timeoutMs);
  if (result === "sat") {
    return `${label}: inner admits a value which outer forbids`;
  }
  if (result !== "unsat") {
    return `${label}: could not prove subsumption (solver ${result}) → denied`;
  }
  return null;
}

function networkScopeViolation(outer: Permission, inner: Permission): string | null {
  if (!inner.network) return null;
  if (!outer.network) return "network: inner uses network but outer forbids it";
  if (outer.network_hosts.length === 0) return null;
  if (inner.network_hosts.length === 0) {
    return `network: inner admits any host but outer restricts to ${JSON.stringify(outer.network_hosts)}`;
  }
  const outerSet = new Set(outer.network_hosts.map((h) => h.toLowerCase()));
  const extra = inner.network_hosts.filter((h) => !outerSet.has(h.toLowerCase()));
  if (extra.length > 0) {
    return `network: inner hosts ${JSON.stringify(extra)} not in outer allowlist ${JSON.stringify(outer.network_hosts)}`;
  }
  return null;
}

async function permissionScopeViolation(outer: Permission, inner: Permission, timeoutMs: number): Promise<string | null> {
  if (inner.shell_allow_decomposition && !outer.shell_allow_decomposition) {
    return "shell_allow_decomposition: inner admits compound shell but outer does not";
  }
  for (const [label, innerP, outerP] of [
    ["file_read", inner.file_read, outer.file_read],
    ["file_write", inner.file_write, outer.file_write],
    ["mcp_allowlist", inner.mcp_allowlist, outer.mcp_allowlist],
  ] as const) {
    const reason = await patternsSubsume(innerP, outerP, label, timeoutMs);
    if (reason !== null) return reason;
  }
  return networkScopeViolation(outer, inner);
}

export function robotCapabilityViolation(outer: Permission, inner: Permission): string | null {
  if (outer.workspace_bounds !== null) {
    if (inner.workspace_bounds === null) {
      return "inner declares no workspace_bounds but outer constrains the workspace (undeclared = unbounded → denied)";
    }
    const [oMin, oMax] = outer.workspace_bounds;
    const [iMin, iMax] = inner.workspace_bounds;
    for (let k = 0; k < 3; k++) {
      if (iMin[k]! < oMin[k]! || iMax[k]! > oMax[k]!) {
        return `inner workspace_bounds exceed outer`;
      }
    }
  }
  for (const axis of ["velocity_limit", "torque_limit"] as const) {
    const oLim = outer[axis];
    if (oLim !== null) {
      const iLim = inner[axis];
      if (iLim === null) return `inner declares no ${axis} but outer caps it (undeclared → denied)`;
      if (iLim > oLim) return `inner ${axis} ${iLim} exceeds outer ${oLim}`;
    }
  }
  for (const [joint, [oLo, oHi]] of Object.entries(outer.joint_limits)) {
    const iRange = inner.joint_limits[joint];
    if (!iRange) return `inner does not bound joint '${joint}' that outer limits (undeclared → denied)`;
    const [iLo, iHi] = iRange;
    if (iLo < oLo || iHi > oHi) return `inner joint '${joint}' range exceeds outer`;
  }
  const freeze = (boxes: Permission["obstacles"]) => new Set(boxes.map((b) => JSON.stringify(b)));
  const outerSet = freeze(outer.obstacles);
  const innerSet = freeze(inner.obstacles);
  const missing = [...outerSet].filter((o) => !innerSet.has(o));
  if (missing.length > 0) {
    return `inner omits ${missing.length} obstacle region(s) the outer forbids (undeclared forbidden region → denied)`;
  }
  return null;
}

function detectInterpreters(perms: Permission): string[] {
  if (!perms.shell) return [];
  return [...new Set(perms.shell_allowlist.filter((n) => SHELL_INTERPRETERS.has(n)))].sort();
}

function shellHeadInAllowlist(cmdVar: string, allowlist: string[]): string {
  if (allowlist.length === 0) return "false";
  const pieces = allowlist.flatMap((head) => [
    `(= ${cmdVar} ${smtStr(head)})`,
    `(str.prefixof ${smtStr(head + " ")} ${cmdVar})`,
  ]);
  return `(or ${pieces.join(" ")})`;
}

function encodeShellAdmission(perms: Permission, cmdVar: string): string {
  if (!perms.shell) return "false";
  const headOk = shellHeadInAllowlist(cmdVar, perms.shell_allowlist);
  const metachars = [";", "|", "&", "`", "<", ">", "\n", "\r"];
  const substrings = ["$("];
  const noMeta = [
    ...metachars.map((ch) => `(not (str.contains ${cmdVar} ${smtStr(ch)}))`),
    ...substrings.map((s) => `(not (str.contains ${cmdVar} ${smtStr(s)}))`),
  ].join(" ");
  return `(and ${headOk} (and ${noMeta}))`;
}

function compileInvariants(
  invariants: Envelope["invariants"],
  scope: Scope,
  env: ReturnType<typeof newEnv>,
  strict: boolean,
  recognizedOpaqueTypes: ReadonlySet<string>
): { term: string; opaque: string[]; strictBlocking: string[] } {
  const opaque: string[] = [];
  const strictBlocking: string[] = [];
  const terms: string[] = [];
  for (const inv of invariants) {
    if (!inv.enforce) continue;
    if (inv.expr === null || inv.expr === undefined) {
      if (strict && !recognizedOpaqueTypes.has(inv.type)) strictBlocking.push(inv.type);
      else opaque.push(inv.type);
      continue;
    }
    let expr = inv.expr;
    if (expr.op === "forall_steps" || expr.op === "exists_step") expr = expr.pred;
    terms.push(compileScalar(expr, scope, env, scope.prefix));
  }
  if (terms.length === 0) return { term: "true", opaque, strictBlocking };
  if (terms.length === 1) return { term: terms[0]!, opaque, strictBlocking };
  return { term: `(and ${terms.join(" ")})`, opaque, strictBlocking };
}

export async function envelopeSubsumes(
  outer: Envelope,
  inner: Envelope,
  timeoutMs: number,
  strict: boolean,
  recognizedOpaqueTypes: ReadonlySet<string>
): Promise<SubsumptionResult> {
  const robotViolation = robotCapabilityViolation(outer.permissions, inner.permissions);
  if (robotViolation !== null) {
    return {
      holds: false,
      unverifiedInvariants: [],
      reasons: [`robot capability subsumption failed (fail-closed): ${robotViolation}`],
    };
  }

  const scopeViolation = await permissionScopeViolation(outer.permissions, inner.permissions, timeoutMs);
  if (scopeViolation !== null) {
    return {
      holds: false,
      unverifiedInvariants: [],
      reasons: [`permission scope subsumption failed (fail-closed): ${scopeViolation}`],
    };
  }

  const policy = outer.shell_interpreter_policy;
  const innerInterpreters = detectInterpreters(inner.permissions);
  const outerInterpreters = detectInterpreters(outer.permissions);
  if (policy === "strict" && innerInterpreters.length > 0) {
    return {
      holds: false,
      unverifiedInvariants: [
        ...new Set([...innerInterpreters, ...outerInterpreters].map((n) => `shell_interpreter:${n}`)),
      ].sort(),
      reasons: [],
    };
  }

  const cmd = "ctx_command";

  // Inner and outer are compiled with INDEPENDENT CompileEnvs (independent
  // soft-node counters, matching the oracle's separate soft_inner/soft_outer
  // lists — soft names are position-keyed within a scope, so inner and outer
  // can legitimately collide on the SAME name at the same position, which is
  // how "both sides declare the same soft predicate" ends up sharing a Z3
  // Bool). Declarations/assumptions are then merged: since both scopes use
  // the SAME "ctx" prefix, a shared path (e.g. "command") always yields the
  // IDENTICAL variable name from either compile, so a plain Set union over
  // "(declare-const …)" lines is exactly the sharing the oracle gets from
  // pre-seeding both `_Scope.vars["ctx__command"]` with the same Z3 var.
  const envInner = newEnv();
  const scopeInner = new Scope("ctx", null);
  envInner.declares.add(`(declare-const ${cmd} String)`);
  const envOuter = newEnv();
  const scopeOuter = new Scope("ctx", null);
  envOuter.declares.add(`(declare-const ${cmd} String)`);

  const innerShell = encodeShellAdmission(inner.permissions, cmd);
  const outerShell = encodeShellAdmission(outer.permissions, cmd);

  const innerCompiled = compileInvariants(inner.invariants, scopeInner, envInner, strict, recognizedOpaqueTypes);
  const outerCompiled = compileInvariants(outer.invariants, scopeOuter, envOuter, strict, recognizedOpaqueTypes);

  if (strict && innerCompiled.strictBlocking.length > 0) {
    return {
      holds: false,
      unverifiedInvariants: [...new Set([...outerCompiled.opaque, ...outerCompiled.strictBlocking])].sort(),
      reasons: innerCompiled.strictBlocking
        .sort()
        .map((t) => `inner invariant '${t}' is opaque and unrecognized; delegation cannot be proven safe under strict mode`),
    };
  }

  const innerAdmits = `(and ${innerShell} ${innerCompiled.term})`;
  const outerAdmits = `(and ${outerShell} ${outerCompiled.term})`;

  // Soft-node polarity: bind every soft node introduced compiling INNER to
  // true (optimistic — inner is maximally permissive). Any soft node that
  // only came from OUTER (not sharing a name with an inner soft node) fails
  // closed: subsumption can't be proven (oracle: Attack C from the v0.11
  // audit — pinning outer-only soft constraints to False was fail-OPEN under
  // negation).
  const softInnerNames = new Set(envInner.soft);
  const softOuterUniqueNames = envOuter.soft.filter((n) => !softInnerNames.has(n));

  if (softOuterUniqueNames.length > 0) {
    return {
      holds: false,
      unverifiedInvariants: [
        ...new Set([
          ...innerCompiled.opaque,
          ...outerCompiled.opaque,
          ...outerCompiled.strictBlocking,
          ...softOuterUniqueNames.map((n) => `soft:${n}`),
        ]),
      ].sort(),
      reasons: [
        `outer has unverifiable soft constraints not shared by inner (${JSON.stringify(
          softOuterUniqueNames.sort()
        )}); cannot prove subsumption (fail-closed)`,
      ],
    };
  }

  const declareLines = [...new Set([...envInner.declares, ...envOuter.declares])].join("\n");
  const assumeLines = [...envInner.assumptions, ...envOuter.assumptions].join("\n");
  const softBindings = envInner.soft.map((n) => `(assert (= ${n} true))`).join("\n");
  const script = [
    declareLines,
    assumeLines,
    softBindings,
    `(assert ${innerAdmits})`,
    `(assert (not ${outerAdmits}))`,
  ]
    .filter(Boolean)
    .join("\n");

  const result = await z3CheckSat(script, timeoutMs);
  if (result === "unknown") {
    throw new VerificationTimeout(`Z3 subsumption check exceeded ${timeoutMs}ms`);
  }

  const interpreterSurface =
    policy !== "allow" ? [...innerInterpreters, ...outerInterpreters].map((n) => `shell_interpreter:${n}`) : [];
  const unverified = [
    ...new Set([
      ...innerCompiled.opaque,
      ...outerCompiled.opaque,
      ...outerCompiled.strictBlocking,
      ...interpreterSurface,
    ]),
  ].sort();

  if (result === "unsat") {
    return { holds: true, unverifiedInvariants: unverified, reasons: [] };
  }
  return { holds: false, unverifiedInvariants: unverified, reasons: [] };
}

File: clients/rust/src/subsumption.rs
//! Port of `opendaisugi/subsumption.py`'s `envelope_subsumes` — the
//! skill-delegation "delegation" stage. Only the `.holds` boolean is
//! wire-relevant (violation comparison is by `(stage, step)`, never by
//! reason), so this port skips reconstructing a Z3 counterexample /
//! `SubsumptionResult` detail payload the oracle builds for logging.
//!
//! Every Z3 call here (like `z3_bridge.rs`) reasons over a fully symbolic
//! scope — `outer ⊨ inner` asks "is there any command Z3 can find that
//! inner would admit and outer would reject", never a concrete plan.

use crate::models::{Envelope, InvariantDecl, Permission};
use crate::predicate::{self, Expr};
use crate::z3_bridge::{self, Scope};
use std::collections::HashSet;

pub struct SubsumptionResult {
    pub holds: bool,
}

fn freeze_box(b: &([f64; 3], [f64; 3])) -> ((u64, u64, u64), (u64, u64, u64)) {
    let f = |x: f64| x.to_bits();
    let (lo, hi) = b;
    ((f(lo[0]), f(lo[1]), f(lo[2])), (f(hi[0]), f(hi[1]), f(hi[2])))
}

/// `_robot_capability_violation` — fail-closed: an outer physical bound the
/// inner exceeds or leaves undeclared blocks delegation. `None` = no
/// violation on this axis (also the correct answer for two non-robot
/// envelopes, where every outer bound is `None`/empty).
fn robot_capability_violation(outer: &Permission, inner: &Permission) -> bool {
    if let Some((o_min, o_max)) = outer.workspace_bounds {
        match inner.workspace_bounds {
            None => return true,
            Some((i_min, i_max)) => {
                for k in 0..3 {
                    if i_min[k] < o_min[k] || i_max[k] > o_max[k] {
                        return true;
                    }
                }
            }
        }
    }
    if let Some(o_lim) = outer.velocity_limit {
        match inner.velocity_limit {
            None => return true,
            Some(i_lim) if i_lim > o_lim => return true,
            _ => {}
        }
    }
    if let Some(o_lim) = outer.torque_limit {
        match inner.torque_limit {
            None => return true,
            Some(i_lim) if i_lim > o_lim => return true,
            _ => {}
        }
    }
    for (joint, (o_lo, o_hi)) in &outer.joint_limits {
        match inner.joint_limits.get(joint) {
            None => return true,
            Some((i_lo, i_hi)) => {
                if i_lo < o_lo || i_hi > o_hi {
                    return true;
                }
            }
        }
    }
    let outer_set: HashSet<_> = outer.obstacles.iter().map(freeze_box).collect();
    let inner_set: HashSet<_> = inner.obstacles.iter().map(freeze_box).collect();
    outer_set.difference(&inner_set).next().is_some()
}

fn glob_unsupported(glob: &str) -> bool {
    if glob == "**" || glob.ends_with("/**") || !glob.contains('*') {
        return false;
    }
    if glob.starts_with('*') && !glob[1..].contains('*') {
        return false;
    }
    glob.matches('*').count() != 1
}

fn smt_string_literal(s: &str) -> String {
    let mut out = String::from("\"");
    for c in s.chars() {
        match c {
            '"' => out.push_str("\"\""),
            '\n' => out.push_str("\\u{a}"),
            '\r' => out.push_str("\\u{d}"),
            other => out.push(other),
        }
    }
    out.push('"');
    out
}

/// `_glob_to_z3` — a permission glob as a Z3 String predicate over `v`.
fn glob_to_z3(glob: &str) -> String {
    if glob == "**" {
        return "true".to_string();
    }
    if let Some(prefix) = glob.strip_suffix("/**") {
        return format!(
            "(or (str.prefixof {} v) (= v {}))",
            smt_string_literal(&format!("{prefix}/")),
            smt_string_literal(prefix)
        );
    }
    if !glob.contains('*') {
        return format!("(= v {})", smt_string_literal(glob));
    }
    if glob.starts_with('*') && !glob[1..].contains('*') {
        return format!("(str.suffixof {} v)", smt_string_literal(&glob[1..]));
    }
    if glob.matches('*').count() == 1 {
        let idx = glob.find('*').unwrap();
        return format!(
            "(and (str.prefixof {} v) (str.suffixof {} v))",
            smt_string_literal(&glob[..idx]),
            smt_string_literal(&glob[idx + 1..])
        );
    }
    "true".to_string() // unsupported shape; caller fail-closes on the OUTER side via glob_unsupported
}

/// `_patterns_subsume` — true iff there's a subsumption VIOLATION on this
/// axis (a witness inner admits that outer forbids, or the check couldn't
/// be proven and so fails closed).
fn patterns_subsume_violates(inner_patterns: &[String], outer_patterns: &[String]) -> Result<bool, String> {
    if inner_patterns.is_empty() {
        return Ok(false); // inner admits nothing on this axis
    }
    if outer_patterns.iter().any(|g| glob_unsupported(g)) {
        return Ok(true); // can't soundly encode outer -> can't prove -> deny
    }
    let mut script = String::new();
    script.push_str("(set-logic ALL)\n(declare-const v String)\n");
    let inner_ok = format!("(or {})", inner_patterns.iter().map(|g| glob_to_z3(g)).collect::<Vec<_>>().join(" "));
    let outer_ok = if outer_patterns.is_empty() {
        "false".to_string()
    } else {
        format!("(or {})", outer_patterns.iter().map(|g| glob_to_z3(g)).collect::<Vec<_>>().join(" "))
    };
    script.push_str(&format!("(assert {inner_ok})\n(assert (not {outer_ok}))\n(check-sat)\n"));
    let result = z3_bridge::check_sat(&script)?;
    Ok(result != "unsat") // sat = witness found; unknown also denies (fail-closed)
}

fn network_scope_violation(outer: &Permission, inner: &Permission) -> bool {
    if !inner.network {
        return false;
    }
    if !outer.network {
        return true;
    }
    if outer.network_hosts.is_empty() {
        return false;
    }
    if inner.network_hosts.is_empty() {
        return true;
    }
    let outer_set: HashSet<String> = outer.network_hosts.iter().map(|h| h.to_lowercase()).collect();
    inner.network_hosts.iter().any(|h| !outer_set.contains(&h.to_lowercase()))
}

fn permission_scope_violates(outer: &Permission, inner: &Permission) -> Result<bool, String> {
    if inner.shell_allow_decomposition && !outer.shell_allow_decomposition {
        return Ok(true);
    }
    for (inner_p, outer_p) in
        [(&inner.file_read, &outer.file_read), (&inner.file_write, &outer.file_write), (&inner.mcp_allowlist, &outer.mcp_allowlist)]
    {
        if patterns_subsume_violates(inner_p, outer_p)? {
            return Ok(true);
        }
    }
    Ok(network_scope_violation(outer, inner))
}

fn detect_interpreters(perms: &Permission) -> Vec<String> {
    if !perms.shell {
        return vec![];
    }
    let mut v: Vec<String> =
        perms.shell_allowlist.iter().filter(|n| crate::models::SHELL_INTERPRETERS.contains(&n.as_str())).cloned().collect();
    v.sort();
    v.dedup();
    v
}

fn shell_head_in_allowlist(allowlist: &[String]) -> String {
    if allowlist.is_empty() {
        return "false".to_string();
    }
    let mut pieces = Vec::new();
    for head in allowlist {
        pieces.push(format!("(= ctx_command {})", smt_string_literal(head)));
        pieces.push(format!("(str.prefixof {} ctx_command)", smt_string_literal(&format!("{head} "))));
    }
    format!("(or {})", pieces.join(" "))
}

/// `_encode_shell_admission`.
fn encode_shell_admission(perms: &Permission) -> String {
    if !perms.shell {
        return "false".to_string();
    }
    let head_ok = shell_head_in_allowlist(&perms.shell_allowlist);
    let metachars = [";", "|", "&", "`", "<", ">", "\n", "\r", "$("];
    let no_meta: Vec<String> =
        metachars.iter().map(|ch| format!("(not (str.contains ctx_command {}))", smt_string_literal(ch))).collect();
    format!("(and {head_ok} (and {}))", no_meta.join(" "))
}

/// `_compile_invariants` — returns (term, strict_blocking_types). Opaque
/// (expr-less) types are tracked in the oracle for `unverified_invariants`
/// reporting only, which is informative-only on the wire, so this port
/// drops that list.
fn compile_invariants(invariants: &[InvariantDecl], scope: &mut Scope, strict: bool) -> Result<(String, Vec<String>), String> {
    let mut strict_blocking = Vec::new();
    let mut terms = Vec::new();
    for inv in invariants {
        if !inv.enforce {
            continue;
        }
        match &inv.expr {
            None => {
                if strict && !crate::z3_checks::RECOGNIZED_OPAQUE_TYPES.contains(&inv.type_.as_str()) {
                    strict_blocking.push(inv.type_.clone());
                }
            }
            Some(v) if v.is_null() => {
                if strict && !crate::z3_checks::RECOGNIZED_OPAQUE_TYPES.contains(&inv.type_.as_str()) {
                    strict_blocking.push(inv.type_.clone());
                }
            }
            Some(v) => {
                let mut expr = predicate::parse_expression(v)?;
                expr = match expr {
                    Expr::ForallSteps { pred } => *pred,
                    Expr::ExistsStep { pred } => *pred,
                    other => other,
                };
                terms.push(z3_bridge::compile_scalar(&expr, scope));
            }
        }
    }
    let term = if terms.is_empty() {
        "true".to_string()
    } else if terms.len() == 1 {
        terms.remove(0)
    } else {
        format!("(and {})", terms.join(" "))
    };
    Ok((term, strict_blocking))
}

/// `envelope_subsumes(outer, inner)` — proves `outer ⊨ inner`. Returns only
/// `.holds`; see the module docstring for why the counterexample detail is
/// dropped.
pub fn envelope_subsumes(outer: &Envelope, inner: &Envelope, strict: bool) -> Result<SubsumptionResult, String> {
    if robot_capability_violation(&outer.permissions, &inner.permissions) {
        return Ok(SubsumptionResult { holds: false });
    }
    if permission_scope_violates(&outer.permissions, &inner.permissions)? {
        return Ok(SubsumptionResult { holds: false });
    }

    let inner_interpreters = detect_interpreters(&inner.permissions);
    if outer.shell_interpreter_policy == "strict" && !inner_interpreters.is_empty() {
        return Ok(SubsumptionResult { holds: false });
    }

    let mut scope_inner = Scope::new("ctx");
    scope_inner.vars.insert("ctx__command".to_string(), ("ctx_command".to_string(), z3_bridge::Sort::Str));
    let mut scope_outer = Scope::new("ctx");
    scope_outer.vars.insert("ctx__command".to_string(), ("ctx_command".to_string(), z3_bridge::Sort::Str));

    let inner_shell = encode_shell_admission(&inner.permissions);
    let outer_shell = encode_shell_admission(&outer.permissions);

    let (inner_inv, inner_strict_blocking) = compile_invariants(&inner.invariants, &mut scope_inner, strict)?;
    let (outer_inv, _outer_strict_blocking) = compile_invariants(&outer.invariants, &mut scope_outer, strict)?;

    if strict && !inner_strict_blocking.is_empty() {
        return Ok(SubsumptionResult { holds: false });
    }

    let outer_soft_unique = scope_outer.soft.iter().any(|n| !scope_inner.soft.contains(n));
    if outer_soft_unique {
        return Ok(SubsumptionResult { holds: false });
    }

    let mut combined = Scope::new("ctx");
    combined.absorb(&scope_inner);
    combined.absorb(&scope_outer);

    let mut script = String::new();
    script.push_str("(set-logic ALL)\n");
    script.push_str(&z3_bridge::declare_block_pub(&combined));
    for name in &scope_inner.soft {
        script.push_str(&format!("(assert (= {name} true))\n"));
    }
    script.push_str(&format!("(assert (and {inner_shell} {inner_inv}))\n"));
    script.push_str(&format!("(assert (not (and {outer_shell} {outer_inv})))\n"));
    script.push_str("(check-sat)\n");

    let result = z3_bridge::check_sat(&script)?;
    Ok(SubsumptionResult { holds: result == "unsat" })
}
